README


  Text::BibTeX
---------------------------------------------------------------------------

  Text::BibTeX is a Perl library for reading, parsing, and processing
  BibTeX files.  It is the Perl half of btOOL, a pair of libraries for
  dealing with BibTeX data.

  Text::BibTeX gives you access to the data at many different levels:
  you may work with BibTeX entries as simple "field -> string"
  mappings, or get at the original form of the data as a list of
  simple values (strings, macros, or numbers) pasted together.  You
  can choose not to impose nany restrictions on the allowed/expected
  entry types or fields, or you can use the structure defined by
  BibTeX 0.99's standard style files, or you can invent your own.

  The library is copiously documented.  After installing the module,
  see the Text::BibTeX man page for a comprehensive introduction to
  the system.  If you wish to dive straight in without regards for
  niceties such as splitting and formatting names, defining or
  imposing database structures, getting access to unprocessed field
  values, and other nifty features, then you can probably get away
  with just reading the Text::BibTeX::Entry man page.
  Comments/criticism of the documentation are welcome.


  In addition to the man pages embedded in each module (and available
  after installation), Greg Ward has written a technical report
  describing btOOL, with the btparse and Text::BibTeX documentation
  included as appendices.  The whole report is just over 100 pages,
  around 45 of which make up the Text::BibTeX documentation (the
  btparse documentation is a further 30 pages).  You can find it at
  the btOOL home page:

    http://starship.python.net/~gward/btOOL/


  INSTALLATION
---------------------------------------------------------------------------

  To install Text::BibTeX you need a recent Perl, an ANSI-compliant C
  compiler and a bunch of Perl modules.

  The build system changed with version 0.40, and has been tested in
  different platforms, ranging from Linux, Mac OS X, Solaris and
  Windows. Unfortunately on this last just the Strawberry Perl
  distribution was a success.

  To build the module use the usual set of commands for a
  Module::Build distribution:

    perl Build.PL
    ./Build
    ./Build test

  And then, as super user, install it

    ./Build install

  Note that if your Perl is installed in a non standard path you can
  end up with libbtparse library in the wrong place. If the tests pass
  successfully, but after installing you can not issue

    perl -MText::BibTeX

  without an error message, then probably libbtparse is not reachable
  by the library loader. In this case you can either copy the library
  to the proper place or add the path to the library in the
  LD_LIBARY_PATH environment variable.


CHANGES
Release 0.66 - 29 October, 2012
-------------------------------
* Fix a segmentation fault with strcat and no string limit.

Release 0.65 - 26 September, 2012
---------------------------------
* Patch to support @ and ~ in names (thanks to Philip Kime)

Release 0.64 - 8 July, 2012
---------------------------
* Patch to expand macro size limit (thanks to Philip Kime)

Release 0.63 - 12 May, 2012
---------------------------
* Patch on what to consider whitespaces (thanks to Philip Kime)

Release 0.62 - 11 January, 2012
-------------------------------
* Releasing stable

Release 0.62_01 - 7 January, 2012
---------------------------------
* Better installation under Windows, and specifically, Win 7.

Release 0.61 - 20 October, 2011
-------------------------------
* Thanks to Philip Kime (again), we have STDERR working on Windows
  (hopefully)
* Make Windows tests no longer ignored.

Release 0.60 - 31 July, 2011
----------------------------
* Fix test in t/output.t

Release 0.59 - 28 July, 2011
----------------------------
* Add comma at the end of each and any line in BibTeX record
  as all modern parsers support them.
* Rewrote some tests in t/output.t

Release 0.58 - 21 June, 2011
----------------------------
* Remove some old documentation about btparse.
* Warn users when installing on non standard library path.

Release 0.57 - 17 June, 2011
----------------------------
* Mark two tests that fail from time to time as TODO for now.

Release 0.56 - 8 June, 2011
---------------------------
* Use File::Temp instead of POSIX. Fixes some Win32 builds.
  Thanks to Alex Gough for reporting this build problem.

Release 0.55 - 25 April, 2011
-----------------------------
* Added some extra checks for lib64 dir.
* Rewrote some code on MyBuilder.pm
* Add soname to library build.

Release 0.54 - 17 April, 2011
-----------------------------
* Applied patch to work under lib64 archs (hopefuly)
 (Guillaume Rousse)

Release 0.53 - 10 March, 2011
-----------------------------
* Working under Cygwin (Philip Kime)
* Changed Windows testing mode (based on Cygwin approach)
  - getting rid of 00_ and zz_ test files. Great!

Release 0.52 - 15 February, 2011
--------------------------------
* More name abbreviation tweeks (Philip Kime)

Release 0.51 - 1 February, 2011
-------------------------------
* Fixed environment during Linux/Unix testing

Release 0.50 - 1 February, 2011
-------------------------------
* Fixed name abbreviation when name has hiffens (Philip Kime)

Release 0.49 - 27 January, 2011
-------------------------------
* Small patch by Philip Kime to enlarge macros buffer size.
* Fixed Build.PL to clean some object files.

Release 0.48 - 23 September, 2010
---------------------------------
* depend on ExtUtils::LibBuilder for library compilation.

Release 0.47 - 18 September, 2010
---------------------------------
* fixed a problem with 160 char being considered whitespace on many unixes.
  Thanks, again, to Philip Kime :)

Release 0.46 - 24 August, 2010
------------------------------
* support entry keys in unicode. Thanks to Philip Kime.

Release 0.45 - 8 June, 2010
---------------------------
* fixed ldconfig call under linux, now only if the user is root.
* fixed library installation path when user specifies different
  installbase during install action.
  (thanks to François for both fixes)

Release 0.44 - 9 May, 2010
--------------------------
* RPath information on link - Thanks to Jens Rehsack
* removed dependency on 'UNIVERSAL' as it is now built-in
  (change for perl 5.12 deprecation)

Release 0.43 - 18 March, 2010
-----------------------------
* small fix on warnings issue
* run ldconfig after installing the .so file under linux

Release 0.42 - 16 March, 2010
-----------------------------
* names with more than two commas are left untouched if protected by
  braces.
* can now deal with names with hyphens and generate the correct BibTeX
  compatible abbreviations.
* can generate initials for protected name parts, like BibTeX can.
* can deal with generating initials for names which contains Unicode
  characters.
* fixed compilation on machines needing alloca.h
* added new README file with more up-to-date information.

Release 0.41 - 14 March, 2010
-----------------------------
* tested under Solaris and FreeBSD.
* fixed compilation/installation under Mac OS X.
* fixed some C code on format-security (thanks to Jerome Quelin).
* fixed build system to check for dependencies.
* added patch from Philip Kime to fix names with commas handling.

Release 0.40 - 7 March, 2010
----------------------------
* this release is very similar to 0.40_3, but now indexable by cpan.

Release 0.40_3 - 18 February, 2010
----------------------------------
* fixed Build.PL with correct build_requirements;
* fixed test to use still not installed library;

Release 0.40_2 - 15 February, 2010
----------------------------------
* merged btparse library into Text::BibTeX code. I am sorry for all
  other languages that might be using this library. For them my
  suggestion is that they make Text::BibTeX as a dependency.
* ported compilation and configuration tools from ExtUtils::MakeMaker
  to Module::Build.
* back on track, trying to make Text::BibTeX work and compile easily
  on main platforms (sorry, for main platforms I assume Linux, Mac OS
  X and Windows running Strawberry Perl).
* Thanks to Philip Kime for continuous poke so I work on this!!
  Created a THANKS file for this purpose.

Release 0.38 - 8 March, 2008
----------------------------
* fixed problem when creating an empty Entry.
  Thanks to Frank Wiegand.

Release 0.37 - 21 Setember, 2006
--------------------------------
* added support for 'conference' key (alias for inproceedings)

Release 0.36 - 24 July, 2005
----------------------------
* fixed url where btparse should be downloaded from. (stupid bug)

Release 0.35 - 23 July, 2005
----------------------------
* changed from 0.33 to 0.35 to maintain relation with btparse.

btparse/src/scan.c

/* parser.dlg -- DLG Description of scanner
 *
 * Generated from: bibtex.g
 *
 * Terence Parr, Will Cohen, and Hank Dietz: 1989-1994
 * Purdue University Electrical Engineering
 * With AHPCRC, University of Minnesota
 * ANTLR Version 1.33
 */

#include <stdio.h>
#define ANTLR_VERSION	133

#define ZZCOL
#define USER_ZZSYN

#include "config.h"
#include "btparse.h"
#include "attrib.h"
#include "lex_auxiliary.h"
#include "error.h"
#include "my_dmalloc.h"

extern char * InputFilename;            /* for zzcr_ast call in pccts/ast.c */
#include "../pccts/antlr.h"
#include "../pccts/ast.h"
#include "tokens.h"
#include "../pccts/dlgdef.h"
LOOKAHEAD
void zzerraction()
{
	(*zzerr)("invalid token");
	zzadvance();
	zzskip();
}
/*
 * D L G tables
 *
 * Generated from: parser.dlg
 *
 * 1989-1994 by  Will Cohen, Terence Parr, and Hank Dietz
 * Purdue University Electrical Engineering
 * DLG Version 1.33
 */

#include "mode.h"



static void act1()
{ 
		NLA = 1;
	}


static void act2()
{ 
		NLA = AT;
		at_sign ();   
	}


static void act3()
{ 
		NLA = 3;
		newline ();   
	}


static void act4()
{ 
		NLA = COMMENT;
		comment ();   
	}


static void act5()
{ 
		NLA = 5;
		zzskip ();   
	}


static void act6()
{ 
		NLA = 6;
		toplevel_junk ();   
	}

static unsigned char shift0[257] = {
  0, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  4, 2, 5, 5, 4, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 4, 5, 5, 5, 5, 3, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 1, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5
};


static void act7()
{ 
		NLA = 1;
	}


static void act8()
{ 
		NLA = 7;
		newline ();   
	}


static void act9()
{ 
		NLA = COMMENT;
		comment ();   
	}


static void act10()
{ 
		NLA = 8;
		zzskip ();   
	}


static void act11()
{ 
		NLA = NUMBER;
	}


static void act12()
{ 
		NLA = NAME;
		name ();   
	}


static void act13()
{ 
		NLA = LBRACE;
		lbrace ();   
	}


static void act14()
{ 
		NLA = RBRACE;
		rbrace ();   
	}


static void act15()
{ 
		NLA = ENTRY_OPEN;
		lparen ();   
	}


static void act16()
{ 
		NLA = ENTRY_CLOSE;
		rparen ();   
	}


static void act17()
{ 
		NLA = EQUALS;
	}


static void act18()
{ 
		NLA = HASH;
	}


static void act19()
{ 
		NLA = COMMA;
	}


static void act20()
{ 
		NLA = 18;
		start_string ('"');   
	}

static unsigned char shift1[257] = {
  0, 14, 14, 14, 14, 14, 14, 14, 14, 14, 
  3, 1, 14, 14, 3, 14, 14, 14, 14, 14, 
  14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 
  14, 14, 14, 3, 5, 13, 11, 5, 2, 5, 
  14, 8, 9, 5, 5, 12, 5, 5, 5, 4, 
  4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 
  5, 5, 10, 5, 5, 14, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 14, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 6, 5, 7, 14, 14, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 
  5, 5, 5, 5, 5, 5, 5
};


static void act21()
{ 
		NLA = 1;
	}


static void act22()
{ 
		NLA = 19;
		check_runaway_string ();   
	}


static void act23()
{ 
		NLA = 20;
		zzreplchar (' '); zzmore ();   
	}


static void act24()
{ 
		NLA = 21;
		open_brace ();   
	}


static void act25()
{ 
		NLA = 22;
		close_brace ();   
	}


static void act26()
{ 
		NLA = 23;
		lparen_in_string ();   
	}


static void act27()
{ 
		NLA = 24;
		rparen_in_string ();   
	}


static void act28()
{ 
		NLA = STRING;
		quote_in_string ();   
	}


static void act29()
{ 
		NLA = 26;
		zzmore ();   
	}

static unsigned char shift2[257] = {
  0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  2, 1, 3, 3, 2, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 8, 3, 3, 3, 3, 
  3, 6, 7, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 9, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 4, 3, 5, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 
  3, 3, 3, 3, 3, 3, 3
};

#define DfaStates	38
typedef unsigned char DfaState;

static DfaState st0[7] = {
  1, 2, 3, 4, 5, 6, 38
};

static DfaState st1[7] = {
  38, 38, 38, 38, 38, 38, 38
};

static DfaState st2[7] = {
  38, 38, 38, 38, 38, 38, 38
};

static DfaState st3[7] = {
  38, 38, 38, 38, 38, 38, 38
};

static DfaState st4[7] = {
  38, 7, 8, 9, 7, 9, 38
};

static DfaState st5[7] = {
  38, 38, 38, 38, 5, 38, 38
};

static DfaState st6[7] = {
  38, 38, 38, 6, 38, 6, 38
};

static DfaState st7[7] = {
  38, 7, 8, 7, 7, 7, 38
};

static DfaState st8[7] = {
  38, 38, 38, 38, 38, 38, 38
};

static DfaState st9[7] = {
  38, 7, 8, 9, 7, 9, 38
};

static DfaState st10[16] = {
  11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 
  21, 22, 23, 24, 38, 38
};

static DfaState st11[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st12[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st13[16] = {
  38, 25, 26, 26, 26, 26, 26, 26, 26, 26, 
  26, 26, 26, 26, 26, 38
};

static DfaState st14[16] = {
  38, 38, 38, 14, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st15[16] = {
  38, 38, 38, 38, 15, 16, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st16[16] = {
  38, 38, 38, 38, 16, 16, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st17[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st18[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st19[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st20[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st21[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st22[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st23[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st24[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st25[16] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38, 38, 38, 38, 38, 38
};

static DfaState st26[16] = {
  38, 25, 26, 26, 26, 26, 26, 26, 26, 26, 
  26, 26, 26, 26, 26, 38
};

static DfaState st27[11] = {
  28, 29, 30, 31, 32, 33, 34, 35, 36, 31, 
  38
};

static DfaState st28[11] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38
};

static DfaState st29[11] = {
  38, 38, 37, 37, 38, 38, 38, 38, 38, 38, 
  38
};

static DfaState st30[11] = {
  38, 38, 31, 31, 38, 38, 38, 38, 38, 31, 
  38
};

static DfaState st31[11] = {
  38, 38, 31, 31, 38, 38, 38, 38, 38, 31, 
  38
};

static DfaState st32[11] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38
};

static DfaState st33[11] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38
};

static DfaState st34[11] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38
};

static DfaState st35[11] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38
};

static DfaState st36[11] = {
  38, 38, 38, 38, 38, 38, 38, 38, 38, 38, 
  38
};

static DfaState st37[11] = {
  38, 38, 37, 37, 38, 38, 38, 38, 38, 38, 
  38
};


DfaState *dfa[38] = {
	st0,
	st1,
	st2,
	st3,
	st4,
	st5,
	st6,
	st7,
	st8,
	st9,
	st10,
	st11,
	st12,
	st13,
	st14,
	st15,
	st16,
	st17,
	st18,
	st19,
	st20,
	st21,
	st22,
	st23,
	st24,
	st25,
	st26,
	st27,
	st28,
	st29,
	st30,
	st31,
	st32,
	st33,
	st34,
	st35,
	st36,
	st37
};


DfaState accepts[39] = {
  0, 1, 2, 3, 6, 5, 6, 0, 4, 6, 
  0, 7, 8, 0, 10, 11, 12, 13, 14, 15, 
  16, 17, 18, 19, 20, 9, 0, 0, 21, 22, 
  23, 29, 24, 25, 26, 27, 28, 22, 0
};

void (*actions[30])() = {
	zzerraction,
	act1,
	act2,
	act3,
	act4,
	act5,
	act6,
	act7,
	act8,
	act9,
	act10,
	act11,
	act12,
	act13,
	act14,
	act15,
	act16,
	act17,
	act18,
	act19,
	act20,
	act21,
	act22,
	act23,
	act24,
	act25,
	act26,
	act27,
	act28,
	act29
};

static DfaState dfa_base[] = {
	0,
	10,
	27
};

static unsigned char *b_class_no[] = {
	shift0,
	shift1,
	shift2
};



#define ZZSHIFT(c) (b_class_no[zzauto][1+c])
#define MAX_MODE 3
#include "../pccts/dlgauto.h"
